src/celeritas/optical/model/MieModel.cc

```diff
@@ -0,0 +1,114 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/model/MieModel.cc
+//---------------------------------------------------------------------------//
+#include "MieModel.hh"
+
+#include <limits>
+
+#include "corecel/Assert.hh"
+#include "corecel/Types.hh"
+#include "corecel/data/CollectionBuilder.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/inp/Grid.hh"
+#include "celeritas/optical/CoreParams.hh"
+#include "celeritas/optical/CoreState.hh"
+#include "celeritas/optical/InteractionApplier.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+#include "celeritas/optical/MieData.hh"
+#include "celeritas/optical/action/ActionLauncher.hh"
+#include "celeritas/optical/action/TrackSlotExecutor.hh"
+#include "celeritas/optical/model/MieExecutor.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Create a model builder from imported data.
+ */
+auto MieModel::make_builder(SPConstImported imported, Input input)
+    -> ModelBuilder
+{
+    CELER_EXPECT(imported);
+    return [imported = std::move(imported),
+            input = std::move(input)](ActionId id) {
+        return std::make_shared<MieModel>(id, imported, input);
+    };
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the model from imported data and imported material parameters.
+ */
+MieModel::MieModel(ActionId id, SPConstImported imported, Input input)
+    : Model(id, "optical-mie", "interact by optical Mie scattering")
+    , imported_(ImportModelClass::mie, std::move(imported))
+{
+    HostVal<MieData> data;
+    CollectionBuilder builder{&data.mie_record};
+
+    for (auto const& mie : input.data)
+    {
+        MieMaterialData record;
+        record.backward_g = mie.backward_g;
+        record.forward_g = mie.forward_g;
+        record.forward_ratio = mie.forward_ratio;
+
+        builder.push_back(record);
+    }
+
+    data_ = CollectionMirror<MieData>{std::move(data)};
+    CELER_ENSURE(data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build the mean free paths for the model.
+ */
+void MieModel::build_mfps(OptMatId mat, MfpBuilder& build) const
+{
+    CELER_EXPECT(mat < imported_.num_materials());
+
+    if (auto const& mfp = imported_.mfp(mat))
+    {
+        build(mfp);
+    }
+    else
+    {
+        build(inp::Grid::from_constant(1e7));
```

Comment:
Verify consistency with input data: if the MFP are there, check g and ratio are in bounds using the records accessible by `this->host_ref().mie_data[mat]`.